Adding an extension method to return the emotion ranked list with name and respective score

andreluizsecco opened the pull request:
Creation of an extension method to return the emotion ranked list with the name and score.
The motivation for this implementation is described in issue #5560.

Sorting criteria used in ranked list: Score is the primary key (descending), and the name is the secondary key.

The FaceDetectionTests class has been modified to test the new feature.

andreluizsecco committed:
Adding an extension method to return the emotion ranked list with name and respective score

azuresdkci commented:
Can one of the admins verify this patch?

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Azure/azure-sdk-for-net?pullRequest=5561) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Azure/azure-sdk-for-net?pullRequest=5561) before we can accept your contribution.<br/><br/>:x: andreluizsecco [sign now](https://cla.opensource.microsoft.com/Azure/azure-sdk-for-net?pullRequest=5561)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Azure/azure-sdk-for-net?pullRequest=5561) it.</sub>

dsgouda commented:
@andreluizsecco please join the azure github org [here](https://aka.ms/azuregithub)

andreluizsecco commented:
@dsgouda I tried, but I got this error
![error](https://user-images.githubusercontent.com/7199850/54957975-09ed8c00-4f33-11e9-9056-5deacd667160.PNG)

dsgouda commented:
@AlexGhiondea Please review this PR

andreluizsecco commented:
@dsgouda please, can you help me to join the azure github org?

AlexGhiondea commented on src/SDKs/CognitiveServices/dataPlane/Vision/Face/Face/Customizations/EmotionExtensions.cs:
```diff
@@ -0,0 +1,34 @@
+            }
+            .OrderByDescending(emotion => emotion.Value)
+            .ThenBy(emotion => emotion.Key)
+            .ToList();
```
Since the method returns `IEnumerable` why call `ToList` here instead of returning the result for `ThenBy`?

andreluizsecco replied:
You're right, I already changed.

AlexGhiondea commented on src/SDKs/CognitiveServices/dataPlane/Vision/Face/Face/Customizations/EmotionExtensions.cs:
```diff
@@ -0,0 +1,34 @@
+    {
+        /// <summary>
+        /// Create a sorted key-value pair of emotions and the corresponding scores, sorted from highest score on down.
+        /// Sorting criteria: Score is the primary key, and the name is the secondary key.
```
I would update the comment to say the secondary key is sorted alphabetically.

andreluizsecco replied:
I agree, I have already updated the comment.

AlexGhiondea reviewed (changes requested):
The change itself looks fine (module a couple of questions)

Is there an issue tracking the need for this?

andreluizsecco committed:
Removing ToList call and updating comment of the method ToRankedList

Changes:
1) Removed ToList call because is not necessary;
2) Updating the method comment to better describe your sorting criteria

andreluizsecco commented:
> The change itself looks fine (module a couple of questions)
> 
> Is there an issue tracking the need for this?

@AlexGhiondea  Yes, I opened this Issue: #5560

andreluizsecco commented:
@AlexGhiondea Please, let me know if any further fix is ​​required.

yangyuan reviewed (commented):
The extension looks good to me.
But there are some necessary changes missing. For example release notes/version bumping.

pakrym commented:
Feels like this needs at least one test that exercises all the intended features: 
set all properties of Emotion to some values and assert you are getting expected sort order for both of them.

andreluizsecco committed:
Adding unit test for emotion ranked list

andreluizsecco commented:
@pakrym That's a good idea! Please, look at my last commit.

andreluizsecco commented:
> 
> 
> The extension looks good to me.
> But there are some necessary changes missing. For example release notes/version bumping.

@yangyuan I don't know how to proceed in this, could you help me?

andreluizsecco commented:
@dsgouda  @AlexGhiondea @yangyuan @pakrym Hey guys, any update on this PR?

andreluizsecco commented:
@dsgouda @AlexGhiondea @yangyuan @pakrym Hello again! Would it be better to create a separate package to add functionality and extensions to Azure SDKs? I do not think we're evolving much here.

andreluizsecco commented:
@dsgouda @AlexGhiondea @yangyuan @pakrym After almost 2 months, will this PR be accepted or not?

AlexGhiondea commented:
At this point I think we are good to merge unless @yangyuan requires additional changes.

andreluizsecco commented:
@yangyuan, any other changes?

AlexGhiondea commented:
@yangyuan any more comments here?

TFR258 commented:
Any update on this PR? It looks like it is ready to be merged.

Thanks,
Tom

andreluizsecco commented:
Today, 7 months ago, I sent this PR :)

AlexGhiondea commented:
Hi @andreluizsecco.

Apologies for the very long delay. If you could rebase this I will get it merged today.

andreluizsecco committed:
rebase

andreluizsecco committed:
Test fix

andreluizsecco commented:
@AlexGhiondea Done!
Thank you.

pakrym commented:
Thank you @andreluizsecco
